### DISCUSSION TITLE
Cpp vs CCpp targets

### DISCUSSION THREAD

**Discussion by stustd** [lf-lang/lingua-franca#997] [Q&A] [answered]
The documentation says:

> NOTE: If you intend to use C++ code or import C++ libraries in the C target, we provide a special [CCpp target](https://github.com/lf-lang/lingua-franca/wiki/Writing-Reactors-in-C#the-ccpp-target) that automatically uses a C++ compiler by default. Alternatively, you might want to look at the native [Cpp](https://github.com/lf-lang/lingua-franca/wiki/Writing-Reactors-in-Cpp) target, built for C++ code if features such as [federated execution](https://github.com/lf-lang/lingua-franca/wiki/Distributed-Execution) are not needed.

I'm interested in the combination that CCpp provides: application code in C++ and federation. However, what's the (recommended) advantage of using the C++ scheduler instead of the C version (where the latter provides both, federation and clock synchronization)?

**Comment by edwardalee** [comment, marked as answer]
In the Cpp target, reactions are written using a C++ style, whereas in the CCpp target, they are written with a C style. Specifically, accessing inputs, state variables, and parameters is via method calls in the Cpp target.  Also, the type system is more flexible in the Cpp target, supporting generics and handling allocation/deallocation in a more intuitive way. Also, the code generated by the Cpp target is more elegant and is object oriented.

I think (hope) the Cpp target will eventually support federated execution, at which point the CCpp target may become less interesting, but I'm not sure about that.

**Reply by stustd** [reply]
Thanks. Indeed, Cpp and federation would be a very nice feature to have! I'll have a look at what it'll take to do that (...although, knowing c++ and LF at the functional level, by no means I'm a low-level expert in both...).

**Reply by cmnrd** [reply]
Yes, federated C++ is on my agenda, but it also requires quite some work and I haven't got a hold of it yet. Feel free to reach out to me if you are interested in getting involved.

What will likely come relatively soon is a prototype version of federated execution in C++ targeting execution on top of ROS2.

**Reply by stustd** [reply]
Thanks for the offer Christian, after I've informed myself in comparing the C and C++ scheduler implementations I'll reach out to you for help.

**Reply by stustd** [reply]
My interest in Lingua Franca (LF) can be summarized by the motto that "A person's non-deterministic decision is not exclusively his/her decision."  IMO it's also the core issue the real-time community is grappling with (e.g. [ROS2, a viable real-time solution?](https://pouya-moh.com/ros-2-a-viable-real-time-solution/)). Can we say that LF's approach conclusively addresses this problem?

**Reply by lhstrh** [reply]
@stustd: could you elaborate a bit on this motto? Is this about control decisions being affected by external factors that are beyond the control of the system engineer? If so, then I think, yes, LF confronts this problem. But, interestingly, it is not a black-and-white issue; there are trade-offs. These trade-offs become more apparent in distributed applications (see [Quantifying and Generalizing the CAP Theorem (preprint)](https://arxiv.org/abs/2109.07771).

**Reply by stustd** [reply]
@lhstrh: indeed, that was the intent of my question, and I'm glad to hear it confirmed. In the terminology of the CA{P,L} paper, indeed, when Availability decreases (i.e. user latency increases), synchronization will break as dependent reactions at some point in time will have to resume without that information; then we're in a (partially) asynchronous, non-deterministic communication world. These effects will have to be accounted for in mission-critical systems, Under (controlled) lab conditions, however, the LF paradigm (with milestone violations traced as errors), provides a rock-solid testing framework.

**Reply by lhstrh** [reply]
What is key here is that Lingua Franca provides guarantees under _quantifiable assumptions_. In centralized coordination, violations of assumptions about observable latency yield a reduction of availability. In decentralized coordination, violations of assumptions about observable latency yield a reduction of consistency.
